### PR TITLE
OIDC federated Logout handled with idToken hint

### PR DESCRIPTION
**Pull request by Chamodii**

### CODE REVIEW

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/model/StateInfoOIDC.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package org.wso2.carbon.identity.application.authenticator.oidc.model;
```
Comment (Chamodii) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/model/StateInfoOIDC.java]: Add licences

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java [hunk 2]
```diff
@@ -428,6 +449,33 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
         }
     }
 
+    @Override
+
+    protected void initiateLogoutRequest(HttpServletRequest request, HttpServletResponse response, AuthenticationContext context) throws LogoutFailedException {
+
+            if (StringUtils.isNotBlank(getLogoutUrl(context.getAuthenticatorProperties()))) {
+                String logoutPage = getLogoutUrl(context.getAuthenticatorProperties());
+                String idTokenHint=((StateInfoOIDC) context.getStateInfo()).getIdTokenHint();
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java]: format

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java [hunk 3]
```diff
@@ -428,6 +449,33 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
+                    logoutPage=logoutPage+"?id_token_hint="+idTokenHint;
+                }
+               String callback=getCallbackUrl(context.getAuthenticatorProperties());
+                if(StringUtils.isNotBlank(callback)){
+                    logoutPage=logoutPage+"&post_logout_redirect_uri="+callback;
+                }
+                if(StringUtils.isNotBlank(sessionID)){
+                    logoutPage=logoutPage+"&state="+sessionID;
+                }
+                try {
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java]: format all from line number 458  to 470

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/model/StateInfoOIDC.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+import org.wso2.carbon.identity.application.authentication.framework.AuthenticatorStateInfo;
+
+public class StateInfoOIDC extends AuthenticatorStateInfo {
+    private String idTokenHint;
+
+    public void setIdTokenHint(String idTokenHint){
+        this.idTokenHint=idTokenHint;
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/model/StateInfoOIDC.java]: format

Comment (piraveena) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/model/StateInfoOIDC.java]: add new line

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/model/StateInfoOIDC.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    }
+
+    public String getIdTokenHint(){
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/model/StateInfoOIDC.java]: add new line after method signature

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/model/StateInfoOIDC.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package org.wso2.carbon.identity.application.authenticator.oidc.model;
+
+import org.wso2.carbon.identity.application.authentication.framework.AuthenticatorStateInfo;
+
+public class StateInfoOIDC extends AuthenticatorStateInfo {
+    private String idTokenHint;
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/model/StateInfoOIDC.java]: add a new line after the class

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java [hunk 7]
```diff
@@ -556,6 +604,7 @@ protected OAuthClientRequest getAccessTokenRequest(AuthenticationContext context
             callbackUrl = getCallbackUrl(authenticatorProperties);
         }
 
+
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java]: remove extra new line

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java [hunk 8]
```diff
@@ -428,6 +449,33 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
         }
     }
 
+    @Override
+
+    protected void initiateLogoutRequest(HttpServletRequest request, HttpServletResponse response, AuthenticationContext context) throws LogoutFailedException {
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java]: format the method signature. break into two lines

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java [hunk 9]
```diff
@@ -134,6 +145,12 @@ protected String getCallbackUrl(Map<String, String> authenticatorProperties) {
         return callbackUrl;
     }
 
+    protected String getLogoutUrl(Map<String, String> authenticatorProperties) {
+
+        return authenticatorProperties.get(OIDCAuthenticatorConstants.OIDC_LOGOUT_URL);
+    }
+
+
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java]: remove extra new line

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java [hunk 10]
```diff
@@ -104,6 +112,9 @@ public boolean canHandle(HttpServletRequest request) {
                 request.getParameter(OIDCAuthenticatorConstants.OAUTH2_ERROR) != null) {
             //if sends error like access_denied
             return true;
+        }else if (request.getParameter(OIDCAuthenticatorConstants.OAUTH2_PARAM_STATE) != null) {
+            //if sends error like access_denied
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java]: add another comment as  handle post logout response

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java [hunk 11]
```diff
@@ -381,6 +396,10 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
                         authenticatorProperties.get(OIDCAuthenticatorConstants.CLIENT_ID));
             }
 
+            StateInfoOIDC stateInfoOIDC=new StateInfoOIDC();
```
Comment (darshanasbg) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java]: Fix formatting

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java [hunk 12]
```diff
@@ -428,6 +447,34 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
         }
     }
 
+    @Override
+
+    protected void initiateLogoutRequest(HttpServletRequest request, HttpServletResponse response,
+                                         AuthenticationContext context) throws LogoutFailedException {
+
+        if (StringUtils.isNotBlank(getLogoutUrl(context.getAuthenticatorProperties()))) {
```
Comment (Chamodii) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java]: Extract the logic into a separate method  isLogoutEnabled

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java [hunk 13]
```diff
@@ -428,6 +447,34 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
+            String logoutPage = getLogoutUrl(context.getAuthenticatorProperties());
```
Comment (darshanasbg) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java]: logoutPage -> logoutURL

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java [hunk 14]
```diff
@@ -428,6 +447,34 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
+            String idTokenHint = ((StateInfoOIDC) context.getStateInfo()).getIdTokenHint();
```
Comment (darshanasbg) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java]: check instanceof

Comment (darshanasbg) [reply components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java]: getIdTokenHint() private method

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java [hunk 15]
```diff
@@ -428,6 +447,34 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
+            String sessionID = context.getContextIdentifier();
+            if (StringUtils.isNotBlank(idTokenHint)) {
+                logoutPage = logoutPage + "?id_token_hint=" + idTokenHint;
+            }
+            String callback = getCallbackUrl(context.getAuthenticatorProperties());
+            if (StringUtils.isNotBlank(callback)) {
+                logoutPage = logoutPage + "&post_logout_redirect_uri=" + callback;
+            }
+            if (StringUtils.isNotBlank(sessionID)) {
```
Comment (Chamodii) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java]: state will not be empty

File: components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java [hunk 16]
```diff
@@ -428,6 +447,34 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
...
+
+        if (StringUtils.isNotBlank(getLogoutUrl(context.getAuthenticatorProperties()))) {
+            String logoutPage = getLogoutUrl(context.getAuthenticatorProperties());
+            String idTokenHint = ((StateInfoOIDC) context.getStateInfo()).getIdTokenHint();
+            String sessionID = context.getContextIdentifier();
+            if (StringUtils.isNotBlank(idTokenHint)) {
+                logoutPage = logoutPage + "?id_token_hint=" + idTokenHint;
+            }
+            String callback = getCallbackUrl(context.getAuthenticatorProperties());
+            if (StringUtils.isNotBlank(callback)) {
```
Comment (Chamodii) [inline components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java]: blank check for callbackurl not needed

**Pull request closed without merging**